src/bin/ion/commands/beta/generate/templates/rust/result.templ

```diff
@@ -0,0 +1,32 @@
+/// Represents serde result
+pub type SerdeResult<T> = Result<T, SerdeError>;
+
+/// Represents an error found during code generation
+#[derive(Debug)]
+pub enum SerdeError {
+    IonError { source: IonError },
+    SerializationError { description: String },
+    DeserializationError { description: String },
+}
```

Comment:
I don't think this should be included in the template. If you ever try to compose code that was generated from different models, you'll have two different `SerdeError`s with the same name, but you'll have to write code to be able to convert form one to the other.

Can you explain the rationale for not using `IonError` and `IonResult`?

Reply:
Option 1 is possible, but not great. We should not require users to take a dependency on `ion-schema-rust` just to use the generated code.

Option 2 is reasonable, though I'd consider limiting the scope of what gets made public. Right now, if you have an `Element` that is an Ion int, and you call `expect_float()`, you will get a `DecodingError`. It's really a sort of type constraint error. Perhaps `DecodingError` is inappropriately overloaded, but using it is a possibility that doesn't introduce meaning other than the meaning it already has.

I think Option 3 is right out. Bad input is to be expected and should be easily recoverable; calling `expect()` can panic, which is not as easy to recover as an `Err`.

I think a 4th option is to add a new (possibly feature-gated) Result/Error type in `ion-rust` that is specific for this case and has a public constructor.

Finally, we could create a new crate called `ion-codegen-runtime-support` (or something like that) with the error type and any other common functionality, but this seems a little heavyweight to me.

@zslayton do you have any thoughts on this?

Reply:
My hope was to have `IonError` mean that `ion-rust` produced an error, and not something in a layer higher up the stack. Making `IonFailure` public means committing to an API, but I'm hopeful that we can get more helpful errors (requiring storing extra context) soon.

>  Right now, if you have an Element that is an Ion int, and you call expect_float(), you will get a DecodingError. It's really a sort of type constraint error.

This is a good observation. I could be convinced that `IonError` should have a `ConstraintError` variant; however, I'd still want ISL validation errors to result in a distinct type.

------

Per our offline conversation, for the time being we're going to generate a `ValidationError` (name tbd) that's used by all of the types produced in the same codegen run. This covers the 80% use case, and someone using code generated from two or more disparate schemas can still unify them into a common type as they bubble up with `?` by implementing `From<ValidationError> for UnifiedError`. We should probably add an option to give `ValidationError` a custom name.

Reply:
I used `SerdeError` as the name here because `read_from` and `write_to` APIs could result into either validation error or `IonError`. And `SerdeError` sounds like a good name to incorporate both of these errors. Although I can change the variant names of `SerdeError` enum to point to the specific error that is either `IonError` or `ValidationError`. What do you guys think? @zslayton @popematt